File: HOT100/2.双指针/42.接雨水.cpp
/*
 * @lc app=leetcode.cn id=42 lang=cpp
 *
 * [42] 接雨水
 */

// @lc code=start
class Solution {
    public:
        int trap(vector<int>& height) {
            int l = 0, r = height.size() - 1;
            int pre_max = -1, suf_max = -1;
            int res = 0;
            while (l <= r)
            {
                pre_max = max(pre_max, height[l]);
                suf_max = max(suf_max, height[r]);
                res += pre_max > suf_max ? suf_max - height[r --] : pre_max - height[l ++];
            }
            return res;
        }
    };
// @lc code=end

class Solution {
    public:
        int trap(vector<int>& height) {
            vector<int> pre_max, suf_max;
            int cur_pre_max = -1;
            for (auto h : height)
            {
                cur_pre_max = max(cur_pre_max, h);
                pre_max.push_back(cur_pre_max);
            }
            int cur_suf_max = -1;
            for (int i = height.size() - 1; i >= 0; i -- )
            {
                cur_suf_max = max(cur_suf_max, height[i]);
                suf_max.push_back(cur_suf_max);
            }
            reverse(suf_max.begin(), suf_max.end());
            int res = 0;
            for (int i = 0; i < height.size(); i ++ )
            {
                res += min(pre_max[i], suf_max[i]) - height[i];
            }
            return res;
        }
    };

File: HOT100/HOT100p4/42.接雨水.py
#
# @lc app=leetcode.cn id=42 lang=python3
#
# [42] 接雨水
#
from mytools import *
# @lc code=start
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        ll, rr = height[0], height[n - 1]
        l, r = 0, n - 1
        res = 0
        while l < r:
            ll = max(ll, height[l])
            rr = max(rr, height[r])
            if height[l] < height[r]:
                res += ll - height[l]; l += 1
            else: res += rr - height[r]; r -= 1
        return res
# @lc code=end

